packages/apidom-ls/src/services/validation/linter-functions.ts

```diff
@@ -1009,9 +1009,33 @@ export const standardLinterfunctions: FunctionItem[] = [
     functionName: 'apilintOpenAPIPathTemplateValid',
     function: (element: Element) => {
       if (isStringElement(element)) {
-        return true;
+        const parseResult = parse(toValue(element));
+        const parts: string[] = [];
+        parseResult.ast.translate(parts);
+
+        const templateExpressions = parts
+          .filter((part) => part[0] === 'template-expression')
+          .map((part) => part[1].slice(1, -1));
```

Comment:
Would it be possible to be more semantic? By `more semantic` I mean not getting outside of the real of ApiDOM (by converting to  POJO). Performance will greatly suffer as Path Items and Operations are the biggest objects. Ideally we want to stay inside ApIDOM all the time and only do minimal necessary  ApiDOM -> POJO conversions.

I've drafted the code that is semantic does minimal conversions and works on top of ApiDOM:

```js
function: (element: Element) => {
      if (isStringElement(element)) {
        const parameterElements: any[] = [];

        // we're treating PathItemElement polymorphically as ObjectElement as this function has to support all OpenAPI versions
        const pathItemElement = (element.parent as MemberElement).value as ObjectElement;
        if (isArrayElement(pathItemElement.get('parameters'))) {
          parameterElements.push([...pathItemElement.get('parameters')]);
        }

        // special case for handling empty PathItemElement
        if (pathItemElement.isEmpty) {
          return true;
        }

        // we're asserting OperationElement polymorphically as this function has to support all OpenAPI versions
        const operationElements = Array.from(
          pathItemElement.filter((elm: Element) => elm.element === 'operation').map((elm) => elm),
        );
        // collect all ParameterElements from all OperationElements
        operationElements.reduce((params, operationElement) => {
          if (isArrayElement(operationElement.get('parameters'))) {
            params.push(Array.from(operationElement.get('parameters')));
          }
          return params;
        }, parameterElements);

        // filter only appropriate ParameterElements
        const allowedLocation = ['path', 'query'];
        const applicableParameterElements = parameterElements.filter((parameterElement) =>
          allowedLocation.includes(toValue(parameterElement.in)),
        );

        // prepare for path template resolution
        const pathTemplateResolveParams = applicableParameterElements.reduce(
          (params, parameterElement) => {
            params[toValue(parameterElement.name)] = 'placeholder'; // eslint-disable-line no-param-reassign
            return params;
          },
          {},
        );
        const pathTemplate = toValue(element);
        const resolvedPathTemplate = resolve(pathTemplate, pathTemplateResolveParams);

        return test(resolvedPathTemplate, { strict: true });
      }
      return true;
    },
```

Have a look at it. It also handles the case of empty `Path Item Object`. The code also abstracts us from `openapi-path-templating` AST and instead we use it's functions.

Reply:
There is still lot of stuff missing. I can think of `parameterElements.push`. Before we push anything to `parameterElements` we must make sure that what does inside is `ObjectElement` and it's `element` property equals `parameter`.

Reply:
Another thing is that I'm not sure this will handle cases where `Reference Object` is used instead of `Parameter Object`. We will know as long as we have a test case for it. It might be possible that @frantuma already did changes so that linting is performed on dereferenced structure.

Reply:
Right. I agree it's better to stay inside ApiDOM. I'll try to refactor my implementation based on this draft and I'll also try to address the other two comments (`parameterElements.push` and `Reference Object`).

Reply:
Couple of more notes:

`(element.parent as MemberElement).value` - this is guaranteed by the refractor layer to always be `PathItemElement` so no asserts are needed here.

`return !test(resolvedPathTemplate, { strict: true });` - this is guaranteed to work because `openapi-path-templating` already supports that in latest version. 

![image](https://github.com/swagger-api/apidom/assets/193286/28fa2bb8-a165-4156-914f-a751aceb5c89)

If I'll think of something else I'll add comments here.

Reply:
Related to `Reference Object` - if we found out that the linting is performed on non-dereferenced structure (which might be the case on 90%)  I can see following options:

### Option 1

Count the number of remaining template expressions and the the count of `Reference Objects`. If the count equals we just return true. It's hacky, but it's better than nothing. The corner case is that two distinct `Reference Objects` might point to the same `Parameter Object`. 

### Option 2

If there is a `Reference Object` within our collected parameters, we just don't do the validation at all and return `true`. 

---

For first iteration I would just do for **Option 2**.